[ADP-3306] Make more HTTP API errors machine-readable.

paweljakubas opened the pull request:
- [x] Convert to machine-readable error: `AssetNotPresent`
- [x] Convert to machine-readable error: `OutputTokenBundleSizeExceedsLimit`  
- [x] Convert to machine-readable error: `OutputTokenQuantityExceedsLimit`
- [x] Convert to machine-readable error: `TransactionAlreadyInLedger`
- [x] Convert to machine-readable error: `WrongEncryptionPassphrase` 

### Issue Number
ADP-3306

paweljakubas committed:
handle AssetNotPresent

paweljakubas committed:
refactor OutputTokenBundleSizeExceedsLimit

paweljakubas committed:
refactor OutputTokenQuantityExceedsLimit

paweljakubas committed:
some cleaning plus using error in integration testing

paweljakubas committed:
refactor TransactionAlreadyInLedger

paweljakubas committed:
refactor WrongEncryptionPassphrase

paweljakubas committed:
cleaning

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs:
```diff
@@ -356,23 +367,32 @@ instance IsServerError ErrMkTransaction where
+                    , pretty (view #quantity e)
+                    , ". Maximum allowable token quantity: "
+                    , pretty (view #quantityMaxBound e)
+                    , "."
```
Suggestion: can we delete this section?

Justification: all the relevant values are already included in the JSON record, so there's no need to repeat them in the error string.

```suggestion
```

paweljakubas replied:
good point

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs:
```diff
@@ -1021,38 +1045,53 @@ instance IsServerError ErrBalanceTxOutputError where
+                    , pretty address
+                    , ". Asset count: "
+                    , pretty assetCount
+                    , "."
```
Similar to https://github.com/cardano-foundation/cardano-wallet/pull/4622#discussion_r1634142800

```suggestion
                    , "outputs."
```

paweljakubas replied:
:+1:

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs:
```diff
@@ -1021,38 +1045,53 @@ instance IsServerError ErrBalanceTxOutputError where
+                    , pretty (show quantity)
+                    , ". Maximum allowable token quantity: "
+                    , pretty (show quantityMaxBound)
+                    , "."
```
Similar to https://github.com/cardano-foundation/cardano-wallet/pull/4622#discussion_r1634142800

```suggestion
```

paweljakubas replied:
:+1:

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Types/Error.hs:
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
 
+data ApiErrorOutputTokenQuantityExceedsLimit = ApiErrorOutputTokenQuantityExceedsLimit
+    { address
+        :: !Text
```
Can we replace this with `ApiAnyAddress`?

```suggestion
        :: !ApiAnyAddress
```

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Types/Error.hs:
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
+    { address
+        :: !Text
+    , policyId
+        :: !Text
```
Request: use `TokenPolicyId` here.

```suggestion
        :: !(ApiT TokenPolicyId)
```

paweljakubas replied:
done

paolino replied:
This seems in conflict with the idea that `ApiT` is only used "outside" . But ok, the wallet is going in one direction the deposit wallet in another :whatever:

jonathanknowles replied:
> This seems in conflict with the idea that `ApiT` is only used "outside" . But ok, the wallet is going in one direction the deposit wallet in another :whatever:

Hi @paolino! I'm curious, what solution would you suggest here?

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Types/Error.hs:
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
+    , policyId
+        :: !Text
+    , assetName
+        :: !Text
```
Request: use `AssetName` here.
```suggestion
        :: !(ApiT AssetName)
```

paweljakubas replied:
done

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Types/Error.hs:
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
+    { address
+        :: !Text
+    , bundleSize
+        :: !Int
```
Request: replace this with `Natural`

Justification: it's impossible to have a bundle that is negatively-sized.

```suggestion
        :: !Natural
```

paweljakubas replied:
done

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Types/Error.hs:
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
+    , assetName
+        :: !Text
+    , quantity
+        :: !Natural
```
Request: use `TokenQuantity` here

```suggestion
        :: !(ApiT TokenQuantity)
```

paweljakubas replied:
done

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Types/Error.hs:
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
+    , quantity
+        :: !Natural
+    , maxQuantity
+        :: !Natural
```
Request: use `TokenQuantity` here

```suggestion
        :: !(ApiT TokenQuantity)
```

paweljakubas replied:
done

jonathanknowles commented on lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/CoinSelections.hs:
```diff
@@ -426,18 +426,17 @@ spec = describe "SHELLEY_COIN_SELECTION" $ do
+                    (toText policyId)
+                    (toText assetName)
+                    (unTokenQuantity excessiveQuantity)
+                    (unTokenQuantity txOutMaxTokenQuantity)
```
```suggestion
                    (apiAddress targetAddress)
                    policyId
                    assetName
                    excessiveQuantity
                    txOutMaxTokenQuantity
```

jonathanknowles commented on lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/CoinSelections.hs:
```diff
@@ -475,12 +474,11 @@ spec = describe "SHELLEY_COIN_SELECTION" $ do
+            decodeErrorInfo sel `Lifted.shouldBe`
+                OutputTokenBundleSizeExceedsLimit
+                ( ApiErrorOutputTokenBundleSizeExceedsLimit
+                    (toText $ apiAddress targetAddress)
```
```suggestion
                    (apiAddress targetAddress)
```

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs:
```diff
@@ -356,23 +367,32 @@ instance IsServerError ErrMkTransaction where
+                , policyId = toText $ view (#asset . #policyId) e
+                , assetName = toText $ view (#asset . #assetName) e
+                , quantity = unTokenQuantity $ view #quantity e
+                , maxQuantity = unTokenQuantity $ view #quantityMaxBound e
```
```suggestion
                { address = view #address e
                , policyId = view (#asset . #policyId) e
                , assetName = view (#asset . #assetName) e
                , quantity = view #quantity e
                , maxQuantity = view #quantityMaxBound e
```

jonathanknowles reviewed (commented):
Hi @paweljakubas 

Thanks for making this PR!

I have two general comments:

- **Suggestion:**
  Can we convert the error strings into _constants_ (i.e., stop building interpolated error strings)? Now that we have JSON objects to carry the variable parts of the error, I think there's no need to repeat this information within the error strings.
  
  _Example: https://github.com/cardano-foundation/cardano-wallet/pull/4622#discussion_r1634142800_
  
- **Request:**
  Within API error record types, can we replace the use of low-level types like `Text` and `Natural` with specialised types such as `ApiT TokenPolicyId`, `ApiT TokenQuantity`, `ApiAnyAddress`? The existing API error types all use specialised types.

  _Example: https://github.com/cardano-foundation/cardano-wallet/pull/4622#discussion_r1634144889_

paweljakubas committed:
responding to review remarks

jonathanknowles committed:
Fix formatting in `Api.Types.Error`.

In this commit, we wrap excessively-long lines to fit within the line
length limit (80 characters).

jonathanknowles committed:
Regenerate `ApiErrorOutputTokenBundleSizeExceedsLimit.json`.

The `bundle_size` field value can never be negative.

jonathanknowles committed:
Regenerate `ApiErrorOutputTokenQuantityExceedsLimit.json`

We can no longer use invalid values for `policy_id` and `asset_name`.

jonathanknowles committed:
Fix formatting in `Api.Http.Server.Error`.

jonathanknowles committed:
Regenerate `ApiError.json`.

The `bundle_size` field value can never be negative.

paweljakubas committed:
[ADP-3306] Fixes for PR #4622. (#4627)

Fixes for PR #4622.

jonathanknowles reviewed (approved):
LGTM! 👍🏻 

I think it makes sense to defer making a decision on using a more dedicated type for the `address` fields for now. Switching to `ApiAddress n`, where `n` is a `NetworkDiscriminant` would require quite a few invasive changes.
